### PR TITLE
Possibility to apply discount on order

### PR DESCRIPTION
**Pull request by korycins**
I want to merge this change because it introduces the possibility to discount the draft orders.

# Impact

* [ ] New migrations
* [ ] New/Updated API fields or mutations
* [ ] Deprecated API fields or mutations
* [ ] Removed API types, fields, or mutations
* [ ] Documentation needs to be updated

# Pull Request Checklist

* [ ] Privileged queries and mutations are guarded by proper permission checks
* [ ] Database queries are optimized and the number of queries is constant
* [ ] Database migration files are up to date
* [ ] The changes are tested
* [ ] GraphQL schema and type definitions are up to date
* [ ] Changes are mentioned in the changelog

### CODE REVIEW

File: saleor/graphql/order/mutations/discount_order.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+    order_line = graphene.Field(
+        OrderLine, description="Order line which has been discounted."
+    )
+    order = graphene.Field(
+        Order, description="Order which is related to the discounted line."
+    )
+
+    class Arguments:
+        order_line_id = graphene.ID(
+            description="ID of a order_line to update price", required=True
```
Comment (maarcingebala) [inline saleor/graphql/order/mutations/discount_order.py:253]:
```suggestion
            description="ID of a order line to update price", required=True
```

File: saleor/graphql/order/mutations/discount_order.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+    order_line = graphene.Field(
+        OrderLine, description="Order line which has removed discount."
+    )
+    order = graphene.Field(
+        Order, description="Order which is related to line which has removed discount."
+    )
+
+    class Arguments:
+        order_line_id = graphene.ID(
+            description="ID of a order_line to remove its discount", required=True
```
Comment (maarcingebala) [inline saleor/graphql/order/mutations/discount_order.py:328]:
```suggestion
            description="ID of a order line to remove its discount", required=True
```

File: saleor/graphql/order/types.py [hunk 3]
```diff
@@ -155,33 +221,42 @@ def resolve_invoice_number(root: models.OrderEvent, _info):
...
-                    break
-            results.append(
-                OrderEventOrderLineObject(
-                    quantity=raw_line["quantity"],
-                    order_line=line_object,
-                    item_name=raw_line["item"],
+            line_pk = entry.get("line_pk", None)
+            if not line_pk:
+                continue
+            line_pks.append(line_pk)
```
Comment (maarcingebala) [inline saleor/graphql/order/types.py:235]: Minor stylistic change but I think this is more intuitive:
```suggestion
            if line_pk:
                line_pks.append(line_pk)
```

File: saleor/graphql/order/types.py [hunk 4]
```diff
@@ -271,6 +353,24 @@ class OrderLine(CountableDjangoObjectType):
...
+        TaxedMoney,
+        description=(
+            "Price of the single item in the order line without applied an order line "
+            "discount."
+        ),
+        required=True,
+    )
+    unit_discount = graphene.Field(
+        Money,
+        description=("The discount applied to the single order line."),
```
Comment (maarcingebala) [inline saleor/graphql/order/types.py:366]: I think we can drop parentheses:
```suggestion
        description="The discount applied to the single order line.",
```

File: saleor/graphql/order/types.py [hunk 5]
```diff
@@ -548,7 +754,7 @@ def resolve_lines(root: models.Order, info):
     @staticmethod
     @permission_required(OrderPermissions.MANAGE_ORDERS)
     def resolve_events(root: models.Order, _info):
-        return root.events.all().order_by("pk")
+        return root.events.prefetch_related("user").all().order_by("pk")
```
Comment (maarcingebala) [inline saleor/graphql/order/types.py:757]: Ideally, the `OrderEvent` type should have a data loader for users, I think the simplest one would do the job:
```
class UserByIdLoader(DataLoader):
    context_key = "user_by_id"

    def batch_load(self, keys):
        users = User.objects.in_bulk(keys)
        return [users.get(user_id) for user_id in keys]
```

File: saleor/order/events.py [hunk 6]
```diff
@@ -469,3 +474,154 @@ def order_note_added_event(*, order: Order, user: UserType, message: str) -> Ord
...
+        "reason": order_discount.reason,
+    }
+    if old_order_discount:
+        discount_parameters["old_value"] = old_order_discount.value
+        discount_parameters["old_value_type"] = old_order_discount.value_type
+        discount_parameters["old_amount_value"] = old_order_discount.amount_value
+    return discount_parameters
+
+
+def order_discount_event(
```
Comment (maarcingebala) [inline saleor/order/events.py:497]: I would consider moving these to a new file `saleor/order/events/discounts.py` but if you don't have time let's skip it for now. Later we could try to refactor other events.

File: saleor/order/utils.py [hunk 7]
```diff
@@ -65,12 +67,51 @@ def decorator(*args, **kwargs):
...
+        amount = order_discount.amount
+
+        if (
+            (
+                discount_type == DiscountValueType.PERCENTAGE
+                or current_total < discount_value
+            )
+            # No need to add new event when new and old amount of discount is the same
+            and amount != current_order_discount.amount
+        ):
```
Comment (maarcingebala) [inline saleor/order/utils.py:107]: Maybe some of these boolean expressions could be assigned to variables? The `if` condition would be easier to read.

File: saleor/order/utils.py [hunk 8]
```diff
@@ -65,12 +67,51 @@ def decorator(*args, **kwargs):
...
             except NotApplicable:
                 discount = zero_money(order.currency)
-            order.discount = discount
-        return func(*args, **kwargs)
+        return func(*args, **kwargs, discount=discount)
 
     return decorator
 
 
+def get_voucher_discount_assigned_to_order(order: Order):
```
Comment (maarcingebala) [inline saleor/order/utils.py:75]: Again, I would consider if we can move all purely discount-related functions to `saleor/order/utils/discounts.py`.

File: CHANGELOG.md [hunk 9]
```diff
@@ -22,6 +23,10 @@ All notable, unreleased changes to this project will be documented in this file.
 - Drop `CHECKOUT_QUANTITY_CHANGED` webhook - #6797 by @d-wysocki
 - Drop deprecated `taxRate` field from `ProductType` - #6795 by @d-wysocki
 - Remove resolving user's location from GeoIP; drop `PaymentInput.billingAddress` input field - #6784 by @maarcingebala
+- Change the payload of the order webhook to handle discounts list, added fields: `Order.discounts`,
+`OrderLine.unit_discount_amount`,`OrderLine.unit_discount_type`, `OrderLine.unit_discount_reason` , remove fields:
+`Order.discount_amount`, `Order.discount_name`, `Order.translated_discount_name`.
+Deprecate API fields `Order.discount`, `Order.discountName`, `Order.translatedDiscountName` - #6874 by @korycins
```
Comment (fowczarek) [inline CHANGELOG.md:29]: I'm not sure that is `Breaking`.

**Review by fowczarek — APPROVED** [review]

File: saleor/checkout/complete_checkout.py [hunk 10]
```diff
@@ -366,6 +365,20 @@ def _create_order(
         status=status,
         channel=checkout.channel,
     )
+    if checkout.discount:
+        # store voucher as a fixed value as it this the simplest solution for now.
+        # This will be solved when we refactor the voucher logic to use .discounts
```
[the remark this answers is not in the document]
Comment (korycins) [reply saleor/checkout/complete_checkout.py:370]: We will include this thread to the task after I will analyze all requirements for this.

File: saleor/graphql/order/types.py (hunk 3 above)
Comment (korycins) [reply saleor/graphql/order/types.py:235]: Good catch. I was suggesting with the old code too much! thanks

File: saleor/graphql/order/types.py (hunk 5 above)
Comment (korycins) [reply saleor/graphql/order/types.py:757]: We already discovered the issue with the missing dataloader here and in OrderEvent type.  I believe @fowczarek already created the tasks for this as the changes require more work to do it properly. . I've added here the prefetch as I discover that we are generating huge amount of queries by calling order events. But the loaders here should be added in separate PR.

**Pull request merged**